src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java

```diff
@@ -1135,37 +1150,48 @@ public int compare(Segment o1, Segment o2) {
 
     @Override
     public void close() throws ElasticsearchException {
-        logger.debug("close now acquire writeLock");
-        try (InternalLock _ = writeLock.acquire()) {
-            logger.debug("close acquired writeLock");
-            if (isClosed.compareAndSet(false, true)) {
+        if (isClosed.get() == false) { // don't acquire the write lock if we are already closed
+            logger.debug("close now acquire writeLock");
+            try (InternalLock _ = writeLock.acquire()) {
+                logger.debug("close now acquired writeLock");
+                closeNoLock();
+            }
+        }
+    }
+
+    /**
+     * Closes the engine without acquiring the write lock. This should only be
+     * called while the write lock is hold or in a disaster condition ie. if the engine
+     * is failed.
+     */
+    private void closeNoLock() throws ElasticsearchException {
+        if (isClosed.compareAndSet(false, true)) {
+            assert rwl.isWriteLockedByCurrentThread() || failEngineLock.isHeldByCurrentThread() : "Either the write lock must be held or the engine must be currently be failing itself";
+            try {
+                this.versionMap.clear();
+                logger.debug("close searcherManager");
                 try {
-                    closedOrFailed = true;
-                    this.versionMap.clear();
-                    logger.debug("close searcherManager");
+                    IOUtils.close(searcherManager);
+                } catch (Throwable t) {
+                    logger.warn("Failed to close SearcherManager", t);
+                }
+                // no need to commit in this case!, we snapshot before we close the shard, so translog and all sync'ed
+                if (indexWriter != null) {
```

Comment:
I think we don't need the indexWriter != null check. It can't  - it's final (which is great)